src/FsCodec.SystemTextJson/UnionConverter.fs

```diff
@@ -25,3 +41,120 @@ module private Union =
+
+        writer.WritePropertyName(discriminator)
+        writer.WriteStringValue(case.Name)
+
+        match fieldInfos with
+        | [| fi |] ->
+            match fieldValues.[0] with
+            | null when options.IgnoreNullValues -> ()
+            | fv ->
+                let element = JsonSerializer.SerializeToElement(fv, options)
+                match element.ValueKind with
+                | JsonValueKind.Object ->
+                    // flatten the object properties into the same one as the discriminator
+                    for prop in element.EnumerateObject() do
+                        prop.WriteTo writer
+                | _ ->
+                    writer.WritePropertyName(fi.Name)
+                    element.WriteTo writer
+        | _ ->
+            for fieldInfo, fieldValue in Seq.zip fieldInfos fieldValues do
+                if fieldValue <> null || not options.IgnoreNullValues then
+                    writer.WritePropertyName(fieldInfo.Name)
+                    JsonSerializer.Serialize(writer, fieldValue, options)
+
+        writer.WriteEndObject()
+
+    override __.Read(reader, t : Type, options) =
+        reader.ValidateTokenType(JsonTokenType.StartObject)
+        use document = JsonDocument.ParseValue &reader
+        let element = document.RootElement
+
+        let targetCaseIndex =
+            let inputCaseNameValue = element.GetProperty discriminator |> string
+            let findCaseNamed x = union.cases |> Array.tryFindIndex (fun case -> case.Name = x)
+            match findCaseNamed inputCaseNameValue, catchAllCase  with
+            | None, None ->
+                sprintf "No case defined for '%s', and no catchAllCase nominated for '%s' on type '%s'"
+                    inputCaseNameValue typeof<UnionConverter<_>>.Name t.FullName |> invalidOp
+            | Some foundIndex, _ -> foundIndex
+            | None, Some catchAllCaseName ->
+                match findCaseNamed catchAllCaseName with
+                | None ->
+                    sprintf "No case defined for '%s', nominated catchAllCase: '%s' not found in type '%s'"
+                        inputCaseNameValue catchAllCaseName t.FullName |> invalidOp
+                | Some foundIndex -> foundIndex
+
+        let targetCaseFields, targetCaseCtor = union.cases.[targetCaseIndex].GetFields(), union.caseConstructor.[targetCaseIndex]
+        targetCaseCtor (Union.mapTargetCaseArgs element options targetCaseFields) :?> 'T
+
+/// Serializes a discriminated union case with a single field that is a
+/// record by flattening the record fields to the same level as the discriminator
+type UnionConverter (defaultDiscriminator : string, defaultCatchAllCase : string option) =
+    inherit Serialization.JsonConverterFactory()
+
+    static let converterType = typedefof<UnionConverter<_>>
+
+    new() = UnionConverter("case", null)
+    new(discriminator: string, catchAllCase: string) = // Compatibility with Newtonsoft UnionConverter constructor
```

Comment:
can this collapse into a primary ctor with 2 optional args which defaults to `"case"` for discriminator ?